lib/ohai/plugins/linux/filesystem2.rb

```diff
@@ -0,0 +1,216 @@
+      next unless entry[:mount]
+      view[entry[:mount]] = Mash.new unless view[entry[:mount]]
+      entry.each do |key, val|
+        next if key == 'mount'
+        view[entry[:mount]][key] = val
+      end
+    end
+    view
+  end
+
+  collect_data(:linux) do
+    fs = Mash.new
+    have_lsblk = File.executable?('/bin/lsblk')
+
+    # Grab filesystem data from df
+    so = shell_out("df -P")
+    so.stdout.lines do |line|
+      case line
+      when /^Filesystem\s+1024-blocks/
+        next
+      when /^(.+?)\s+(\d+)\s+(\d+)\s+(\d+)\s+(\d+\%)\s+(.+)$/
+        key = "#{$1},#{$6}"
+        fs[key] = Mash.new
+        fs[key][:device] = $1
+        fs[key][:kb_size] = $2
+        fs[key][:kb_used] = $3
+        fs[key][:kb_available] = $4
+        fs[key][:percent_used] = $5
+        fs[key][:mount] = $6
+      end
+    end
+    
+    # Grab filesystem inode data from df
+    so = shell_out("df -iP")
+    so.stdout.lines do |line|
+      case line
+      when /^Filesystem\s+Inodes/
+        next
+      when /^(.+?)\s+(\d+)\s+(\d+)\s+(\d+)\s+(\d+\%)\s+(.+)$/
+        key = "#{$1},#{$6}"
+        fs[key] ||= Mash.new
+        fs[key][:device] = $1
+        fs[key][:total_inodes] = $2
+        fs[key][:inodes_used] = $3
+        fs[key][:inodes_available] = $4
+        fs[key][:inodes_percent_used] = $5
+        fs[key][:mount] = $6
+      end
+    end
+
+    # Grab mount information from /bin/mount
+    so = shell_out("mount")
+    so.stdout.lines do |line|
+      if line =~ /^(.+?) on (.+?) type (.+?) \((.+?)\)$/
+        key = "#{$1},#{$2}"
+        fs[key] = Mash.new unless fs.has_key?(key)
+        fs[key][:device] = $1
+        fs[key][:mount] = $2
+        fs[key][:fs_type] = $3
+        fs[key][:mount_options] = $4.split(",")
+      end
+    end
+
+    have_lsblk = File.exist?('/bin/lsblk')
+    if have_lsblk = File.exist?('/bin/lsblk')
+      cmd = 'lsblk -n -P -o NAME,UUID,LABEL,FSTYPE'
+    else
+      cmd = '"blkid'
+    end
+
+    so = shell_out(cmd)
+    so.stdout.lines do |line|
+      parsed = parse_line(line, have_lsblk)
+      next if parsed.nil?
+      # lsblk lists each device once, so we need to update all entries
+      # in the hash that are related to this device
+      keys_to_update = []
+      fs.each_key do |key|
+        keys_to_update << key if key.start_with?("#{dev},")
+      end
+        
+      if keys_to_update.empty?
+        key = "#{dev},"
+        fs[key] = Mash.new
+        fs[key][:device] = dev
+        keys_to_update << key
+      end
+
+      keys_to_update.each do |key|
+        [:fs_type, :uuid, :label].each do |subkey|
+          if parsed[subkey] && !parsed[subkey].empty?
+            fs[key][subkey] = parsed[subkey]
+          end
+        end
+      end
+    end
+
+    # Grab any missing mount information from /proc/mounts
+    if File.exist?('/proc/mounts')
+      mounts = ''
+      # Due to https://tickets.opscode.com/browse/OHAI-196
+      # we have to non-block read dev files. Ew.
```

Comment:
That can block - we must call `read_nonblock()` - this code is copied wholesale from the original filesystem plugin, fwiw.